### PR TITLE
sql: granular detection of non-immutable constant casts

### PR DESCRIPTION
**Pull request by RaduBerinde**
#### opt: improve tests for parsing timestamps

Add more `sem/tree` unit tests for parsing time-related types. In
particular:
 - relative timestamps ('now', 'tomorrow')
 - TZ variants
 - timestamps that include timezones for DDate, DTime
 - cases where the result depends on the current time and/or timezone.

In a future change, the parsing functions will also return a flag
indicating if the result depends on the context or not; these tests
will be updated to check that flag as well.

Release note: None

#### pgdate: add WithoutTimezone variants

The way we parse DTimestamp (and other non-TZ variants) is convoluted:
we initially parse it just like a DTimestampTZ, and then we strip away
the timezone (adjusting it so the timestamp looks the same as before,
except with UTC location). This means that we incorporate the local
timezone, even though it is inconsequential.

This change adds WithoutTimezone API variants to `pgdate` which avoids
using the local timezone. This will allow extending the parsing APIs
to indicate whether the result depends on the current time or
timezone.

Release note: None

#### sql: use dummy ParseTimeContext during type-checking

We are not supposed to use any context-dependent parsing results
during type-checking. As such, use a dummy ParseTimeContext instead of
having SemaContext implement the interface.

Release note: None

#### sql: granular detection of non-immutable constant casts

Interpreting a timestamp (and similar types) can depend on the current
timezone or even the current time. A recent change prevented these
casts from happening during type-checking.

Unfortunately, this includes a lot of common cases which aren't
actually context-dependent. The most important are dates and
timestamps without time zone (except rare cases like 'now' or
'tomorrow'). In these cases, the type-checked expressions won't seem
constant and won't be able to be used for partitioning, index
predicates, etc.

This change improves the parsing functions to return a
`dependsOnContext` boolean. When this flag is false, the result is
immutable and is replaced during type-checking.

The volatility of the version of date_trunc that returns a TIMESTAMPTZ
is corrected to Stable.

Release note: None

### CODE REVIEW

**Comment by rohany** [comment]
The approach looks right here, but I'll defer to @otan for the actual time details.

File: pkg/sql/sem/tree/type_check.go [hunk 1]
```diff
@@ -204,19 +203,6 @@ func (sc *SemaContext) isUnresolvedPlaceholder(expr Expr) bool {
 	return sc.Placeholders.IsUnresolvedPlaceholder(expr)
 }
 
-// GetLocation returns the session timezone.
-func (sc *SemaContext) GetLocation() *time.Location {
-	if sc == nil || sc.Location == nil || *sc.Location == nil {
-		return time.UTC
-	}
-	return *sc.Location
```
Comment (otan) [inline pkg/sql/sem/tree/type_check.go:212]: can we get rid of the struct methods here?

File: pkg/util/timeutil/pgdate/parsing.go [hunk 2]
```diff
@@ -156,6 +194,33 @@ func ParseTimestamp(now time.Time, mode ParseMode, s string) (time.Time, error)
...
+// ParseTimestampWithoutTimezone converts a string into a timestamp, stripping
+// away any timezone information. Any specified timezone is inconsequential. The
+// returned time always has UTC location.
+//
+// For example, all these inputs return 2020-06-26 01:02:03 +0000 UTC:
+//   - '2020-06-26 01:02:03';
+//   - '2020-06-26 01:02:03+04';
+//   - 'now', if the local local time (in the current timezone) is
+//     2020-06-26 01:02:03. Note that this does not represent the same time.Time
+//     instant, but the one that "reads" the same in UTC.
```
Comment (otan) [inline pkg/util/timeutil/pgdate/parsing.go:227]: s/instant/instance/ ?

**Review by otan — APPROVED** [review]
LGTM (minor comments)

File: pkg/util/timeutil/pgdate/parsing.go (hunk 2 above)
Comment (RaduBerinde) [reply pkg/util/timeutil/pgdate/parsing.go:227]: no, like "time instant"

File: pkg/sql/sem/tree/type_check.go (hunk 1 above)
Comment (RaduBerinde) [reply pkg/sql/sem/tree/type_check.go:212]: sorry, not sure what you mean?

File: pkg/util/timeutil/pgdate/parsing.go (hunk 2 above)
Comment (otan) [reply pkg/util/timeutil/pgdate/parsing.go:227]: ah ok!

File: pkg/sql/sem/tree/type_check.go (hunk 1 above)
Comment (otan) [reply pkg/sql/sem/tree/type_check.go:212]: does the `Location` variable on `semaCtx` need to exist still?

Comment (RaduBerinde) [reply pkg/sql/sem/tree/type_check.go:212]: Ah.. good point, removed.

**Pull request merged**